### ISSUE TITLE
Get number of arguments in Lua function

### ISSUE THREAD

**Issue by NickThissen** [NLua/NLua#220]
I am executing a simple Lua script from a C# application. The script must have a function called 'Execute' (which will be invoked by my application). I want to know how many arguments this function accepts. If the function accepts 2 arguments, I pass the two relevant arguments to the script. If it has 4 arguments, I pass another set of arguments. If there is any other number of arguments I want to notify the user their script is invalid.

I cannot find any way to count the number of arguments in a function. I managed to find the function itself so I can check if it exists, but there seems to be no reference to the arguments anywhere.

	var state = new NLua.Lua();
	state.DoString(source);

	var function = state.GetFunction("Execute");
	if (function != null)
	{
		// TODO: Check number of arguments??
		
		return true; // valid
	}
	else 
	{
		return false; // invalid
	}

Is it possible to get the number of arguments in some way?

**Comment by NickThissen** [comment]
Thanks, but I think you misunderstood the question.

I am letting the user of my application write a Lua script. I am then executing that script from my application, which is written in C#, hence I use NLua to compile and execute it.

The Lua script should have a function called `Execute` which accepts exactly a certain number of arguments. If the number of arguments is wrong, I want to notify the user that their script is invalid.

Even further, my application understands two types of scripts. One "simple" type that accepts 2 arguments, and one "advanced" type that accepts 4 arguments. I want to be able to detect if the script that the user has written is "simple" with 2 arguments or "advanced" with 4 arguments. If there are any other number of arguments, I show a message that it's invalid. Then I can determine which arguments to pass depending on if there are 2 or 4 arguments in the Execute function.

To clarify, I need to know **in my C# code after compiling the Lua** if the script has a function called `Execute` with either 2 or 4 arguments, nothing more and nothing less.

Example: this is **C# code**:

    var state = new NLua.Lua();
    state.DoString(source);

    var function = state.GetFunction("Execute");
    if (function != null)
    {
        // TODO: Check number of arguments??
        int argumentCount = function.Arguments.Count; // This does not exist, what can I use instead?

        if (argumentCount == 2)
            return "Valid simple script";
        else if (argumentCount == 4)
            return "Valid advanced script";
        else
            return "Invalid number of arguments";
    }
    
    return "Invalid script, Execute function not found";

**Comment by HoraceBury** [comment]
To be honest, you are trying to provide C# overloading to a scripting language (no, I don't like calling Lua that, but so many do!) which does not have an overloading feature.

The flexibility of variable parameter signatures in languages like JavaScript and Lua gives them a certain power. C# and other strongly typed languages have a different power.

In your situation I would declare in the documentation that the user-developer needs to accept a given list of parameters and that the location in the parameter set defines which value will be received. Therefore, if you're passing in 6 variables, each variable will have a given location and cannot be interpreted in any other way.

Yes, in some implementations, like Corona SDK, functions accept optional parameters at, say, the start of the parameter list. Thinking of a case in point, the `display.newCircle( [parent,] x, y, radius )` for example. Here, the first parameter is optional. This is not implemented by the calling code but the implementatiom within `newCircle `function itself.

In short, I recommend passing in all the parameters that you can provide and the user-developer's function should make the decision as which values to use.

This is the basis of a programmatic 'contract'... two pieces of code do not and should not know anything about each others' internal workings but the contract - which does not change - defines how they talk to each other.

**Comment by NickThissen** [comment]
I'm not trying to provide optional parameters, I simply want the ability to notify the user that their script is valid or not.

You mention the programmatic 'contract' at the end. That's exactly what I have: the documentation says the script must have either 2 or 4 parameters (but not 1, or 3). If the user makes a script with 2 or 4 parameters, it works fine. If the user makes a script with 3 parameters, it still works fine, he'll just be missing the 4th parameter.

The point is, I want to "enforce" this contract by **telling the user** that he made a mistake in his script. I literally just want the ability to pop up a message that states "Hey, you have 3 arguments, this is probably not what you want".

Right now, I can't find any way to do that because I can't count the number of arguments. 

I am now "parsing" the code myself, simply checking for the line with the Execute function and counting the number of commas between the parentheses. This kind of works but it's obviously not ideal. It would be great if the actual script parser could tell me the number of arguments that a function accepts. If that's not possible, fine, I'll resort to my own crude method.

**Comment by HoraceBury** [comment]
Unfortunately, the problem you face is that Lua (as with JavaScript) does not 'enforce' contracts in the same way that C# does. What you're trying to do is provide a helper/validator for the developer, which is totally understandable.

Lua itself does not have the facility to evaluate functions and their parameter lists, AFAIK.

I think the way you're doing it now is probably the only route. Maybe take a look at the deep Lua documentation and post on stackoverflow.

**Comment by HoraceBury** [comment]
I just did a quick google and found this:

http://stackoverflow.com/questions/3097209/lua-get-the-list-of-parameter-names-of-a-function-from-outside-the-function

Let me know if it helps.

**Comment by NickThissen** [comment]
This may be completely unrelated, but since you keep mentioning Javascript, I am doing this exact task for Javascript scripts, using Jint as the compiler/parser:

    var engine = new Engine(options => options.AllowClr()).Execute(source);

    var exec = engine.Global.GetOwnProperty("execute");
    
    if (exec?.Value != null)
    {
        // Count number of arguments here:
        var count = ((ScriptFunctionInstance) exec.Value.AsObject()).FormalParameters.Length;
        if (count != 2 && count != 4)
        {
            result.HasErrors = true;
            result.Errors.Add(new ScriptError
            {
                Message = "Script must contain a function called 'execute' with two or four parameters."
            });
        }
        else
        {
            result.HasErrors = false;
            result.CompiledScript = new JavascriptCompiledScript(engine, count == 2);
        }
    }
    else
    {
        result.HasErrors = true;
        result.Errors.Add(new ScriptError {Message = "Script must contain a function called 'execute'."});
    }

Sure, the code is not so pretty and I have to kind of dig down into the depths of the code, but it works.


I will take a look at your link. I have thought about trying to get the number of arguments from within Lua itself, by simply running the script and inspecting what happens (or looking at the environment table). But I don't want to simply execute the script because it might have side effects that the user doesn't want to happen unless he explicitly runs the script.
